Add validation of smt parse trees to construct smt_responset instances.

thomasspriggs opened the pull request:
This PR adds validation of smt parse trees to construct smt_responset instances.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

thomasspriggs committed:
Define interface for smt response validation

The `response_or_errort` class is a template, so that it can be reused
for smt_termt and `smt_sortt` sub trees of the SMT responses, whilst
maintaining strong compile time type checking.

The `UNIMPLEMENTED` macro in the `validate_smt_response` function
will be replaced with an implementation in the following commits of this
PR.

thomasspriggs committed:
Add `smt_responset` debug printer to tests

To facilitate debugging of tests and tracking down causes of test
failures.

thomasspriggs committed:
Move smt2irep testing utilities into `unit/testing-utils`

So they can be reused in multiple test `.cpp` files.

thomasspriggs committed:
Add check sat response validation

thomasspriggs committed:
Add validation of SMT success response

thomasspriggs committed:
Implement error messaging for invalid SMT responses

thomasspriggs committed:
Implement validation of `unsupported` smt2 response

thomasspriggs committed:
Add template for node construction with validation propagation

For each non-leaf node in the `smt_responset` tree, the same logic is
required where the child nodes are validated and we construct the
non-leaf node if there are no errors, or collect and propagate the
errors upwards if there are errors validating the child nodes.

The template in this commit abstracts this logic out, so that it doesn't
need to be duplicated for each type of node we construct.

thomasspriggs committed:
Factor out `print_parse_tree` function

This is currently implemented using `pretty`, but this function is used
forward to replace with an implementation specific to our use case which
is more easily readable by users of CBMC.

thomasspriggs committed:
Document naming scheme used for response validation functions

thomasspriggs committed:
Add SMT get-value response validation for bool sorts

thomasspriggs committed:
Add support for validating get-value responses with bit vectors

The template parameter is specified for mapping to upper case characters
to fix a warning compiling on Ubuntu 18. This warning was related to
`std::toupper` being declared with the deprecated `throw ()` in the
standard library and the meaning of this changing between C++ versions.

thomasspriggs committed:
Add more get-value response validation unit tests

NlightNFotis commented on src/solvers/smt2_incremental/smt_response_validation.cpp:
```diff
@@ -37,5 +37,14 @@ template class response_or_errort<smt_responset>;
 
 response_or_errort<smt_responset> validate_smt_response(const irept &parse_tree)
 {
+  if(parse_tree.id() == "sat")
```
⛏️ Why not `switch (parse_tree.id()) { case "sat": }`. Might make it a bit more readable?

thomasspriggs replied:
Because you can't switch on a string in C++.

NlightNFotis replied:
😱

TGWDB replied:
But you can switch on a/the first character... so if you want a big nested tree of case statements... ;)

NlightNFotis reviewed (commented):
LGTM overall 👍🏻

TGWDB commented on src/solvers/smt2_incremental/smt_response_validation.cpp:
```diff
@@ -0,0 +1,327 @@
+  if(parse_tree.get_sub().size() > 2)
+  {
+    return {response_or_errort<smt_responset>{
+      "Error response has multiple error messages."}};
```
Can we do something nicer here? Maybe gather all the error messages into a multiline string?

thomasspriggs replied:
I have done something nicer here. The original parse tree is now included in this error message, as instances where the messages are sub trees could also fall into this case.

thomasspriggs committed:
Include original error for invalid error error message

So that it is still possible to see the error message which was being
reported.
